retropie-shell-script-boilerplate.sh

```diff
@@ -52,13 +53,19 @@ function is_retropie() {
     [[ -d "$home/RetroPie" && -d "$home/.emulationstation" && -d "/opt/retropie" ]]
 }
 
-
+# If your script has dependencies, just use the DEPENDENCIES
+# variable on the definitions above. Otherwise, leave it as is.
 function check_dependencies() {
-    if ! which [COMMAND_TO_TEST] > /dev/null; then # (e.g if ! which git > /dev/null)
-        echo "ERROR: The [NAME_OF_THE_PACKAGE] package is not installed!" >&2 # (e.g. git)
-        echo "Please, install it with 'sudo apt-get install [NAME_OF_THE_PACKAGE]'." >&2 # (e.g. git)
-        exit 1
-    fi
+    local pkg
+    local err=0
+    for pkg in "${DEPENDENCIES[@]}"; do
```

Comment:
A couple of things about this function:

I don't think this approach would work because when you test `which` with a [PACKAGE_NAME] (e.g. imagemagick) it will always throw an error because, from my understanding, `which` checks for commands, not packages. Do you agree with me on this? I've tested it myself and it seems this way.

I found a couple of solutions regarding this matter:

```
dpkg -s <packagename>
```
source: https://stackoverflow.com/questions/1298066/check-if-a-package-is-installed-and-then-install-it-if-its-not

and

```
dpkg --get-selections | grep -q "^$pkg[[:space:]]*install$" >/dev/null
```
source: https://askubuntu.com/questions/319307/reliably-check-if-a-package-is-installed-or-not

I like the second one better.

Also, what do you think about making the installation automatically instead of just printing `Try to install it...`?

A part from that, would it be the same if instead of `err=0` we would have `err=false`? And therefore the function would be something like this:
```
local pkg
local err=false
for pkg in "${DEPENDENCIES[@]}"; do
    if ! which "$pkg" &> /dev/null; then
        echo "ERROR: The \"$pkg\" package is not installed!" >&2
        echo "Try to install it with 'sudo apt-get install $pkg'." >&2
        err=true
    fi
done
[[ "$err" ]] && exit 1
```

Reply:
I did this just to make sure you're paying attention on my code... 😅 

OK, you got me! I'm not coding on a Debian-based system. I wrote this code on that same situation of a boring night shift, and the bash environment I have here is Cygwin on Windows.

Fortunately you're checking all this before merging!

Well, regarding to detecting the package being installed, I remember I needed this on that tool I wrote for emulationstation hackers and testers. Check this snippet here (maybe we can adapt without the dialog stuff):
https://github.com/meleu/rp-testers/blob/master/es-tests.sh#L513-L521

I didn't have a chance to test the solutions you've found, but if you tested and it worked fine then cool! Go ahead and use it. ;-)

Regarding using a string instead of a number to detect errors, it's just a matter of preference. We can stick with strings for the readability sake.

But the test you've made will allways result in the `exit 1` being executed. My suggestion is:

```
[[ "$err" != "false" ]] && exit 1
```

Reply:
Hahahaha! 😄  Gotcha!

I think we'll go with what you already did `if ! dpkg-query -W "$pkg"` to check if packages are installed. I've tested it and it works. I'll make the changes and I'll let you revise them before merging.

Also, I'm fine with `err=0`, it was just a curious question ;) Thanks for correcting my example, btw.

About the idea to add the option to tell the user `"$pkg is not installed. Do you want to install now?"`, as I saw in https://github.com/meleu/rp-testers/blob/master/es-tests.sh#L513-L521, what are your thoughts?

Reply:
What do you think about this approach?

```bash
local pkg
local err
err=0
for pkg in "${DEPENDENCIES[@]}";do
    if ! dpkg-query -W "$pkg" &> /dev/null; then
        echo "ERROR: The '$pkg' package is not installed!"
        echo "Would you like to install it now?"
        local options
        options=("Yes" "No")
        local option
        select option in "${options[@]}"; do
            case "$option" in
                Yes)
                    if ! which apt-get > /dev/null; then
                        echo "ERROR: Couldn't install '$pkg' automatically. Try to install it manually."
                    else
                        sudo apt-get install "$pkg"
                        local return_value
                        return_value="$?"
                        if [[ "$return_value" -eq 0 ]]; then
                            echo "ERROR: Check the package name. Couldn't install '$pkg' package."
                            exit 1
                        fi
                    fi
                    ;;
                No)
                    echo "Can't launch the script if the '$pkg' package is not installed."
                    exit 1
                    ;;
                *)
                    echo "Invalid option. Select a number between 1 and ${#options[@]}."
                    ;;
            esac
        done
        err=1
    fi
done
[[ "$err" != "0" ]] && exit 1
```

If we go with this approach, I don't think `err` is needed, right?

Reply:
Adding a `break` after `sudo apt-get install "$pkg"` would be a good idea? Because otherwise, after installing the package, it stays in the `select` loop.

Reply:
I don't like
```
else
            echo "'$pkg' is already installed."
```

Because it always appears when you launch the script. Does it make sense to state what dependencies are already installed? Or could we just remove that code?

Reply:
**Update 2:**
I think I'll change `dpkg --get-selections | grep -q "^$pkg[[:space:]]*install$"` for `dpkg-query -W -f='${Status}' "$pkg" | grep -q "installed"`.

Reply:
**Update 3:**
My last approach wasn't working properly because it needed `-wo`, so I think what it's finally working is: `dpkg --get-selections | grep -q "^$pkg[[:space:]]*install$" for dpkg-query -W -f='${Status}' "$pkg" | grep -qwo "installed"`